### PR TITLE
Optimisation in finding factors

### PR DESCRIPTION
**Pull request by ccp92**
I've modified the factors method which creates an array of the factors.
As all factors come in pairs, and no factor will ever be larger than the square root of the number, I can reduce greatly the number of loops through each that are required.

### CODE REVIEW

File: lib/rec_one.rb [hunk 1]
```diff
@@ -18,13 +20,19 @@ def validation(first_number, last_number)
...
   end
 
-  def factors(first_number)
-    (1..first_number).select { |integer| (first_number % integer).zero? }
+  def factors(number)
+    factors = []
+    (1..number**0.5).each do |possible_factor|
+      if number % possible_factor == 0
+        factors << possible_factor
+        factors << number/possible_factor unless number/possible_factor == possible_factor
```
Comment (Renny4Real) [inline lib/rec_one.rb]: Maybe replace (number/possible_factor == possible_factor) with (number == possible_factor**2) makes it slightly shorter and more readable.

File: lib/rec_one.rb [hunk 2]
```diff
@@ -18,13 +20,19 @@ def validation(first_number, last_number)
...
-    sum_of_squares = factors_squared.reduce(:+)
-    results.push([number, sum_of_squares]) if integer?(Math.sqrt(sum_of_squares))
+    sum_of_squares = factors(number).map! { |factor| factor**2 }.inject(:+)
+    results.push([number, sum_of_squares]) if integer?(sum_of_squares**0.5)
   end
 
-  def factors(first_number)
-    (1..first_number).select { |integer| (first_number % integer).zero? }
+  def factors(number)
+    factors = []
```
Comment (Renny4Real) [inline lib/rec_one.rb]: Factors = Array.new maybe?

File: lib/rec_one.rb [hunk 3]
```diff
@@ -18,13 +20,19 @@ def validation(first_number, last_number)
...
+    (1..number**0.5).each do |possible_factor|
+      if number % possible_factor == 0
```
[the remark this answers is not in the document]
Comment (ccp92) [reply lib/rec_one.rb]: I always forget about next. I'll implement that now.

File: lib/rec_one.rb (hunk 1 above)
Comment (ccp92) [reply lib/rec_one.rb]: That definitely improves the readability and length. Thanks.

File: lib/rec_one.rb (hunk 2 above)
Comment (ccp92) [reply lib/rec_one.rb]: I had a look at this stack overflow that mentions that they're functionally the same:
https://stackoverflow.com/questions/4601652/ruby-array-creation-array-new-vs

As [] is shorter than Array.new, I think I'll stick the existing implementation. Good suggestion though, thanks!

**Pull request merged**